JITServer profiling infra for CallGraph entries

mpirvu opened the pull request:
For virtual invokes, interface invokes, checkcast and instanceof bytecodes the IProfiler profiles 3 different target RAMClasses in a data structure called `CallSiteProfileInfo`. We will refer to them as slots. The server cannot cache the RAMClasses because they are valid for just one client. Because the SharedProfileCache is supposed to work with many clients, we need to convert the RAMClasses into something else that is client agnostic. In this implementation we will use the ClassRecord from the JITServer AOT cache.

If the server determined that the profiling information received from the client is better than the one it currently has in the SharedProfileCache (JITServerSharedProfileCache::compareBytecodeProfiles(sharedProfileSummary, clientProfileSummary);) then it will proceed to store this info into the shared profile repository by calling `ClientSessionData::storeBytecodeProfileInSharedRepository()`. Conversely, if the server determines that the shared profile info is better than the client profile info, it's going to load this shared info into the per-client IProfiler cache by calling
`ClientSessionData::loadBytecodeDataFromSharedProfileCache()`.

High level sequence of actions:

1. `ClientSessionData::storeBytecodeProfileInSharedRepository()`. 
1.1 The server deserializes the profiling information received from the client and creates `TR_IPBytecodeHashTableEntry` entries using stack memory. The `TR_IPBCDataCallGraph` entries are put into a different vector (called cgEntries) because they need special treatment.
1.2 For cgEntries we collect all the unique RAMClasses that these entries refer to and memorize to which entry and slot they refer to.
1.3 We then convert the unique RAMClasses into AOT ClassRecords by using `getClassRecord(ramClass, ...)`. Some of these conversions may not succeed because the server may not have cached that specific RAMClass.
1.4 The server creates a vector with uncachedRAMClasses and sends an `AOTCache_getROMClassBatch` message to the client.
1.5 The client will `packRemoteROMClassInfo()` and respond with ClassInfos for all missing classes. Note that if arrays are involved, the client may also send (unsolicited) information about the base component classes of those arrays.
1.6 The server will cache the information related to the missing classes: ` JITServerHelpers::cacheRemoteROMClassBatch(...)`
1.7 The server converts the missing RAMClasses into AOT ClassRecords by using `getClassRecord(ramClass,...)` like before. We expect for all such conversions to succeed this time.
1.8 Now that it performed all the RAMClass-->ClassRecord conversions, the server will patch all the profiling entries that are going to be stored in the shared profile cache, replacing the RAMClasses with ClassRecords.
1.9 The cgEntries are added to the set of other profiling entries 
1.10 The server adds the profiling entries to the shared repository by calling `_sharedProfileCache->addBytecodeData(entries,...)`. They will be cloned using memory that persists throughout the lifetime of the server.

2. `ClientSessionData::loadBytecodeDataFromSharedProfileCache()` 
2.1 For the method of interest, the server clones all profiling entries using either client persistent memory (if the information is to be stored in the per-client profile cache) or heap-memory (if the information is to be stored in the per-complation profile cache). This cloning is done with `methodEntry->cloneBytecodeData(trMemory, stable, newEntries, cgEntries);` Pointers to CallGraph entries are also copied into a `cgEntries` vector because they need special processing.
2.2 For cgEntries the server attempts to convert the ClassRecords from the profiling slots, back into RAMClasses specific to the client that performs the compilation. To help with conversion, the clientSession keeps a hashmap called `_classRecordMap`
2.3 ClassRecords that could not be converted into RAMClass with the help of `_classRecordMap` are collected into a vector called `uniqueUncachedClassRecords` (we eliminate duplicates).
2.4 The server will send a message to the client to perform the desired conversions, but the client also needs to know some record information that should help with deserialization. Thus, the server collects all class serialization records and class loader serialization records that the client might need. When building this list, the server uses the `_aotCacheKnownIds` map, to avoid sending information that the client already has.
2.5 The server sends a `AOTCache_getRAMClassFromClassRecordBatch` message with all the class IDs that need to be converted to RAMClasses and all the supporting serialization records.
2.6 The client caches all the serialization records: `deserializer->cacheRecords((const uint8_t *)recordsStr.data(), recordsStr.size(), ...)`
2.7 For requested classIDs the client performs the conversion into a RAMClass: deserializer->getRAMClass(id, comp, wasReset);
2.8 The server receives the response and updates its view of records that the client knows about (_aotCacheKnownIds)
2.9 For each of the entries in `uniqueUncachedClassRecords` we cache the corresponding RAMClass (obtained from the client) into `_classRecordMap` for future use. We also patch-up our CallGraph profiling entries that could not be fixed-up before.
2.10 The server stores the fixed-up entries into the per-client or per-compilation profile cache. If there is any error before this step the server needs to free the persistent memory it might have used for profiling entries in step 2.1

mpirvu commented:
jenkins compile all jdk8,jdk21

mpirvu commented:
jenkins test sanity plinuxjit,xlinuxjit,zlinuxjit,alinux64jit jdk17

mpirvu commented:
@dsouzai @vijaysun-omr  All tests have passed. This PR is ready for merging (after a review).

vijaysun-omr commented:
Regarding this step 
```
For each of the entries in uniqueUncachedClassRecords we cache the corresponding RAMClass (obtained from the client) into _classRecordMap for future use. We also patch-up our CallGraph profiling entries that could not be fixed-up before.
```

Is it possible that the conversion step fails to find a RAM Class at some (new) client that tried to obtain this information from the server ?

mpirvu commented:
> Is it possible that the conversion step fails to find a RAM Class at some (new) client that tried to obtain this information from the server ?

Yes, there are some such failures, mostly classes that have "lambda" or "proxy" in their name.

vijaysun-omr commented:
Thanks, I know we have some issues on occasions correlating the cached profiling information with RAM classes even without JIT server (i.e. just when using the shared classes cache). Are the issues here with the JIT server expected to be roughly in the same cases that we were unable to correlate without the JIT server ? i.e. are there any JIT server specific issues that might end up failing this correlation ?

mpirvu commented:
> Are the issues here with the JIT server expected to be roughly in the same cases that we were unable to correlate without the JIT server ?

That's my expectation. I will investigate this and compare with the SCC case.

dsouzai reviewed (commented):
Overall looks good to me. Had some questions and minor changes. The one likely major change (replacing the three vectors with a single container) doesn't have to be done in this PR if it's going to be a lot of work, but we should at least note it down so that it can be improved in a future PR.

mpirvu commented on runtime/compiler/runtime/JITClientSession.cpp:
```diff
@@ -650,49 +659,339 @@ ClientSessionData::loadBytecodeDataFromSharedProfileCache(J9Method *method, bool
+
+            for (size_t i = 0; i < uniqueUncachedClassRecords.size(); ++i)
+               {
+               auto classRecord = uniqueUncachedClassRecords[i];
+               auto loaderRecord = classRecord->classLoaderRecord();
```
I don't think so. We take the same approach when storing classes in the SCC.
This information is always used with a profile guard (as far as I know) so it is safe to use.

mpirvu commented on runtime/compiler/runtime/JITClientSession.cpp:
```diff
@@ -650,49 +659,339 @@ ClientSessionData::loadBytecodeDataFromSharedProfileCache(J9Method *method, bool
+         }
+      // Validate cgEntries against the profiling entries sent by the client.
+      // We want to make sure that the j9class pointers that we materialized are valid.
+      if (!ipdata.empty() && !cgEntries.empty())
```
Previously there was an assert in there. Now that I replaced that with a vlog message it makes sense to move the test on `TR_VerboseJITServerSharedProfileDetails` up.

mpirvu commented on runtime/compiler/runtime/JITClientSession.cpp:
```diff
@@ -650,49 +659,339 @@ ClientSessionData::loadBytecodeDataFromSharedProfileCache(J9Method *method, bool
+            for (const auto &clientCgEntry : clientCgEntries)
+               {
+               if  (cgEntry->getPC() == clientCgEntry->getPC())
+                  {
```
Done.

mpirvu commented on runtime/compiler/runtime/JITClientSession.cpp:
```diff
@@ -774,17 +1071,160 @@ ClientSessionData::storeBytecodeProfileInSharedRepository(TR_OpaqueMethodBlock *
+         stream->write(JITServer::MessageType::AOTCache_getROMClassBatch, uncachedRAMClasses);
+         auto recv = stream->read<std::vector<JITServerHelpers::ClassInfoTuple>, std::vector<J9Class *>>();
+         auto &classInfoTuples = std::get<0>(recv);
+         auto &uncachedBases = std::get<1>(recv); // This should be empty because uncachedRAMClasses includes the base components for arrays
+         // The classInfoTuples sent by the client also contains the base components that the client
+         // discovered (added at the end), but the 'uncached' vector only contains what the server requested.
+         // We need to add the supplemental base components that the client discovered.
```
I added line 1179 just in case, but I can replace it with an assert. `uncachedBases` should be empty because above (where it says: "// TODO: do we really need to process the base component separately?") we already send the base components in the request. I could have avoided that and let the client discover the base components, at which point line 1179 is needed.

mpirvu replied:
I changed my mind on this one. It's possible that we deal with a class array that the server does not know about. It will ask the client and the client will send ClassInfo both about the array class and the base component class (if it thinks the server does not have it). So in some cases uncachedBases can have something in it.

mpirvu commented:
I addressed all the comments from the review except for the use of a set. I used a vector because when the number of elements is very small a linear search through the vector is faster than a hashtable.

dsouzai commented:
Not sure why the ECA is failing, but I'm unable to log into the Eclipse, so maybe it's related to the issues from the past few days.

dsouzai commented:
jenkins test sanity plinuxjit,xlinuxjit,zlinuxjit,alinux64jit,xlinux jdk17

mpirvu commented:
sanity.functional fails cmdLineTester_jfrEvents_0 in -Xint mode
zlinux openjdk with JITServer had a timeout for jdk_util_1
xlinux openjdk without JITServer failed jdk_lang_j9_1
```
00:21:56  ACTION: main -- Failed. Execution failed: `main' threw exception: java.lang.RuntimeException: Bad e.callStack[1].getMethodName(): expected: log actual: <unknown method>
```
zlinux sanity.functional with JITServer still in progress.

mpirvu commented:
I had a bug in the latest round of changes to address review comments. Fixed with the last forced push.

mpirvu commented:
jenkins test sanity plinuxjit,xlinuxjit,zlinuxjit,alinux64jit jdk17

mpirvu commented:
The only failures are the JFR tests at -Xint so this PR is ready to be merged.
